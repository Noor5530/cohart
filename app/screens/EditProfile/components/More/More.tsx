import React, { useEffect, useState } from "react";
import { View, TouchableOpacity, Image, Text } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import { ActivityIndicator } from "react-native-paper";
import moment from "moment";

import { Input } from "utils/uiPrimitives";
import AppState from "models/reducers";
import RemoveIcon from "assets/ic_remove_text_2.png";
import WarningIcon from "assets/ic_warning_blue.png";
import HideIcon from "assets/ic_hide.png";
import ShowIcon from "assets/ic_show.png";
import SampleAvatar from "assets/sample-avatar.jpeg";
import userProfileMore from "services/userProfileMore";
import styles from "./style";
import { logOutRequest } from "store/actions/userActions";
import { LogOutRequestEnum } from "models/actions/user";

interface IProps {
  closeCollapsibleSection(): void;
  setSubmitForm(): void;
}

const getArrayDataFromString = (data: any) => {
  if (typeof data == "string") {
    return data.split(",").map((item) => item.trim());
  }
  return data;
};

const More: React.FC<IProps> = ({
  closeCollapsibleSection,
  setSubmitForm,
}: IProps) => {
  const dispatch = useDispatch();
  const user = useSelector((state: AppState) => state.user);
  const [submitting, setSubmitting] = useState(false);

  const affiliations = getArrayDataFromString(user.affiliations);
  const seenMeAt = getArrayDataFromString(user.seen_me_at);

  const [affiliation, setAffiliation] = useState(
    affiliations ? [...affiliations, ""] : [""]
  );

  const [meetMeData, setMeetMeData] = useState(
    seenMeAt ? [...seenMeAt, ""] : [""]
  );

  const [currently, setCurrently] = useState(
    user.currently ? user.currently : ""
  );

  const [isShow, setIsShow] = useState(user?.invited_by_toggle || false);

  useEffect(() => {
    setAffiliation(affiliations ? [...affiliations, ""] : [""]);
    setMeetMeData(seenMeAt ? [...seenMeAt, ""] : [""]);
    setCurrently(user.currently ? user.currently : "");
    setIsShow(user?.invited_by_toggle || false);

    return () => {
      setAffiliation([""]);
      setMeetMeData([""]);
      setCurrently("");
      setIsShow(user?.invited_by_toggle || false);
    };
  }, [user]);

  const toggleShowInvite = () => setIsShow(!isShow);

  const removeAffiliation = (position: number) => () =>
    setAffiliation((prev) => prev.filter((_, index) => index !== position));

  const removeMeetMe = (position: number) => () =>
    setMeetMeData((prev) => prev.filter((_, index) => index !== position));

  const onSubmit = async () => {
    setSubmitting(true);

    const filterAffiliations = affiliation.filter((item) => {
      return item !== "" && item.trim() != "";
    });

    const filterSeenMe = meetMeData.filter((item) => {
      return item !== "" && item.trim() != "";
    });

    const payload = {
      id: user._id,
      affiliations: filterAffiliations,
      seen_me_at: filterSeenMe,
      currently,
      invited_by_toggle: isShow,
    };

    setAffiliation([...filterAffiliations, ""]);
    setMeetMeData([...filterSeenMe, ""]);

    try {
      await userProfileMore(payload);
    } catch (error: any) {
      if (error?.response?.status == 401) {
        dispatch(logOutRequest(LogOutRequestEnum.tokenExpire));
      }
    }

    setSubmitForm((prev) =>
      prev.map((_, index) => (index === 3 ? true : false))
    );
    closeCollapsibleSection();
    setSubmitting(false);
  };

  console.log("[More] affiliations ==> " + JSON.stringify(affiliation));
  console.log("[More] seenMeAt ==> " + JSON.stringify(seenMeAt));

  return (
    <View style={styles.container}>
      <View style={styles.affiliation}>
        <Text style={styles.label}>Affiliation:</Text>
        {affiliation.map((place, index) => {
          return (
            <View key={index} style={styles.inputWithIcon}>
              <Input
                maxLength={25}
                style={styles.inputStyle}
                value={place}
                accessibilityLabel="Add another affiliation"
                placeholderTextColor="#7e8d98"
                placeholder={"Add another affiliation"}
                onChangeText={(text) => {
                  if (text.trim().length == 0 && text == " ") {
                    return;
                  }
                  const temp = [...affiliation];
                  temp[index] = text.substring(0, 25);
                  if (index == affiliation?.length - 1) {
                    setAffiliation([...temp, ""]);
                  } else {
                    setAffiliation(temp);
                  }
                }}
              />
              {index !== affiliation.length - 1 && (
                <TouchableOpacity
                  testID={`removeAffiliation-${index}`}
                  style={styles.removeBtn}
                  onPress={removeAffiliation(index)}
                >
                  <Image source={RemoveIcon} style={styles.icon} />
                </TouchableOpacity>
              )}
            </View>
          );
        })}
      </View>

      <View style={styles.meetMe}>
        <Text style={styles.label}>You Might Have Seen Me At:</Text>
        {meetMeData.map((place, index) => {
          return (
            <View key={index} style={styles.inputWithIcon}>
              <Input
                maxLength={25}
                accessibilityLabel="Add more entry"
                style={styles.inputStyle}
                value={place}
                placeholder={"Add more entry"}
                placeholderTextColor="#7e8d98"
                onChangeText={(text) => {
                  if (text.trim().length == 0 && text == " ") {
                    return;
                  }
                  const temp = [...meetMeData];
                  temp[index] = text.substring(0, 25);
                  if (index == meetMeData?.length - 1) {
                    if (meetMeData?.length === 8) {
                      setMeetMeData(temp);
                      return;
                    }
                    setMeetMeData([...temp, ""]);
                  } else {
                    setMeetMeData(temp);
                  }
                }}
              />
              {index !== meetMeData.length - 1 && (
                <TouchableOpacity
                  testID={`removeEntry-${index}`}
                  onPress={removeMeetMe(index)}
                  style={styles.removeBtn}
                >
                  <Image source={RemoveIcon} style={styles.icon} />
                </TouchableOpacity>
              )}
            </View>
          );
        })}
        {meetMeData.length === 8 && (
          <View style={styles.meetMeWarning}>
            <Image source={WarningIcon} />
            <Text style={styles.warningText}>
              You have reached the maximum number of entries
            </Text>
          </View>
        )}
      </View>

      <View style={styles.currently}>
        <Text style={styles.label}>Currently</Text>
        <View style={styles.inputWithIcon}>
          <Input
            maxLength={25}
            testID="currentlyInput"
            style={styles.inputStyle}
            value={currently}
            placeholder={"Share what youâ€™re up to"}
            onChangeText={(text) => {
              if (text.trim().length == 0 && text == " ") {
                return;
              }
              setCurrently(text);
            }}
          />
          {currently.length !== 0 && (
            <TouchableOpacity
              onPress={() => setCurrently("")}
              style={styles.removeBtn}
            >
              <Image source={RemoveIcon} style={styles.icon} />
            </TouchableOpacity>
          )}
        </View>
      </View>

      <View style={styles.labelRow}>
        <Text style={styles.label}>INVITED BY</Text>
        <TouchableOpacity
          style={styles.buttonToggle}
          onPress={toggleShowInvite}
        >
          <Image
            style={styles.toggleStyle}
            source={isShow ? ShowIcon : HideIcon}
          />
        </TouchableOpacity>
      </View>
      <View>
        <View style={styles.inviteRow} opacity={isShow ? 1 : 0.3}>
          <Image
            source={
              user.referral_by?.cover_image
                ? { uri: user.referral_by?.cover_image }
                : SampleAvatar
            }
            style={styles.avatarItem}
          />
          <View style={styles.userInfo}>
            <Text style={styles.name}>{user?.referral_by?.full_name}</Text>
            <Text style={styles.date}>
              {moment(user.referral_by?.last_active).format("DD MMMM YYYY")}
            </Text>
          </View>
        </View>
      </View>

      <TouchableOpacity
        testID="buttonSubmit"
        style={styles.button}
        onPress={onSubmit}
      >
        {submitting ? (
          <ActivityIndicator color="black" size="small" />
        ) : (
          <Text style={styles.buttonText}>{"I'M DONE"}</Text>
        )}
      </TouchableOpacity>
    </View>
  );
};

export default More;
