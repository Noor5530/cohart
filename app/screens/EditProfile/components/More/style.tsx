import { appFonts } from "components/text";
import { StyleSheet } from "react-native";
import { widthPercentageToDP as wp } from "react-native-responsive-screen";

const styles = StyleSheet.create({
  icon: {
    width: 15,
    height: 15,
  },
  toggleStyle: {
    height: 17
  },
  inputWithIcon: {
    marginTop: 5,
    flexDirection: "row",
    width: "100%",
  },
  removeBtn: {
    position: "absolute",
    right: -10,
    paddingRight: 15,
    top: 10,
    width: 70,
    height: '100%',
    alignItems: 'flex-end'
  },
  container: {
    paddingLeft: 16,
    paddingRight: 16,
  },
  label: {
    marginLeft: 7,
    fontSize: 12,
    fontFamily: appFonts.InterRegular,
  },
  inputStyle: {
    marginBottom: 5,
    paddingLeft: 6,
    borderWidth: 0,
    borderBottomWidth: 1,
    width: "100%",
    fontSize: 20,
  },
  meetMeWarning: {
    flexDirection: "row",
    alignItems: "center",
    marginTop: 10,
  },
  warningText: {
    color: "#0000FF",
    marginLeft: 5,
    fontSize: 12,
    fontFamily: appFonts.InterRegular,
  },
  affiliation: {
    marginBottom: 30,
  },
  meetMe: {
    marginBottom: 30,
  },
  currently: {
    marginBottom: 30,
  },
  labelRow: {
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
  },
  avatarItem: {
    width: 52,
    height: 52,
    borderRadius: 26,
    marginRight: 11,
  },
  inviteRow: {
    flexDirection: "row",
    alignItems: "center",
    marginTop: 15,
  },
  name: {
    fontSize: 16,
    fontFamily: appFonts.InterRegular,
    fontWeight: "500",
  },
  date: {
    fontSize: 10,
    color: "black",
    fontFamily: appFonts.InterRegular,
  },
  userInfo: {
    justifyContent: "center",
  },
  button: {
    width: wp(30),
    height: 43,
    borderWidth: 1,
    borderColor: "black",
    justifyContent: "center",
    alignItems: "center",
    borderRadius: 4,
    alignSelf: "center",
    marginTop: 20,
    marginBottom: 20,
  },
  buttonText: {
    fontFamily: appFonts.InterBold,
    fontSize: 18,
  },
  buttonToggle: {
    width: 70,
    alignItems: 'flex-end',
    height: 40,
    justifyContent: 'center'
  }
});

export default styles;
