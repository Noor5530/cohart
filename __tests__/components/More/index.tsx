import "react-native";
import React from "react";
import More from "../../../app/screens/EditProfile/components/More/More";
import renderer from "react-test-renderer";
// import { render, fireEvent } from "@testing-library/react-native";

jest.useFakeTimers();

jest.mock("react-redux", () => ({
  useDispatch: () => {},
  useSelector: () => ({
    user: {
      affiliations: [],
      referral_by: {
        cover_image: "hnnj",
      },
    },
  }),
}));

jest.mock("react-native-paper/src/components/ActivityIndicator", () => {
  return {
    __esModule: true,
    A: true,
    namedExport: jest.fn(),
    default: "mockedDefaultExport",
  };
});

test("renders correctly", () => {
  const tree = renderer
    .create(<More closeCollapsibleSection={jest.fn} setSubmitForm={jest.fn} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

// test("test affiliation input", () => {
//   const { getAllByA11yLabel } = render(
//     <More closeCollapsibleSection={jest.fn} setSubmitForm={jest.fn} />
//   );
//   const answerInputs = getAllByA11yLabel("Add another affiliation");
//   fireEvent.changeText(answerInputs[0], "affiliation1");

//   const expectNumberAffiliation = getAllByA11yLabel("Add another affiliation");

//   expect(expectNumberAffiliation.length).toEqual(2);
// });

// test("test meetMeAt input", () => {
//   const { getAllByA11yLabel } = render(
//     <More closeCollapsibleSection={jest.fn} setSubmitForm={jest.fn} />
//   );
//   const answerInputs = getAllByA11yLabel("Add more entry");
//   fireEvent.changeText(answerInputs[0], "entry1");

//   const expectNumberOfEntry = getAllByA11yLabel("Add more entry");

//   expect(expectNumberOfEntry.length).toEqual(2);
// });

// test("remove affiliation", () => {
//   const { getAllByA11yLabel, getByTestId } = render(
//     <More closeCollapsibleSection={jest.fn} setSubmitForm={jest.fn} />
//   );
//   const answerInputs = getAllByA11yLabel("Add another affiliation");
//   fireEvent.changeText(answerInputs[0], "affiliation1");

//   const expectNumberAffiliation = getAllByA11yLabel("Add another affiliation");
//   fireEvent.changeText(expectNumberAffiliation[1], "affiliation2");
//   const button = getByTestId(`removeAffiliation-1`);
//   fireEvent.press(button);

//   expect(expectNumberAffiliation.length).toEqual(2);
// });

// test("remove meetMeAt", () => {
//   const { getAllByA11yLabel, getByTestId } = render(
//     <More closeCollapsibleSection={jest.fn} setSubmitForm={jest.fn} />
//   );
//   const answerInputs = getAllByA11yLabel("Add more entry");
//   fireEvent.changeText(answerInputs[0], "entry1");

//   const expectNumberOfEntry = getAllByA11yLabel("Add more entry");
//   fireEvent.changeText(expectNumberOfEntry[1], "entry2");
//   const button = getByTestId(`removeEntry-1`);
//   fireEvent.press(button);

//   expect(expectNumberOfEntry.length).toEqual(2);
// });

