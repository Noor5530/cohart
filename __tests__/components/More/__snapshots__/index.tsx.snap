// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "paddingLeft": 16,
      "paddingRight": 16,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Inter-Regular",
          "fontSize": 12,
          "marginLeft": 7,
        }
      }
    >
      Affiliation:
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 5,
          "width": "100%",
        }
      }
    >
      <TextInput
        accessibilityLabel="Add another affiliation"
        allowFontScaling={true}
        maxLength={25}
        onChangeText={[Function]}
        placeholder="Add another affiliation"
        placeholderTextColor="#7e8d98"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 8,
              "borderWidth": 1,
              "paddingHorizontal": 12,
              "paddingVertical": 4,
            },
            Object {
              "borderBottomWidth": 1,
              "borderWidth": 0,
              "fontSize": 20,
              "marginBottom": 5,
              "paddingLeft": 6,
              "width": "100%",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Object {
        "marginBottom": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Inter-Regular",
          "fontSize": 12,
          "marginLeft": 7,
        }
      }
    >
      You Might Have Seen Me At:
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 5,
          "width": "100%",
        }
      }
    >
      <TextInput
        accessibilityLabel="Add more entry"
        allowFontScaling={true}
        maxLength={25}
        onChangeText={[Function]}
        placeholder="Add more entry"
        placeholderTextColor="#7e8d98"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 8,
              "borderWidth": 1,
              "paddingHorizontal": 12,
              "paddingVertical": 4,
            },
            Object {
              "borderBottomWidth": 1,
              "borderWidth": 0,
              "fontSize": 20,
              "marginBottom": 5,
              "paddingLeft": 6,
              "width": "100%",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Object {
        "marginBottom": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Inter-Regular",
          "fontSize": 12,
          "marginLeft": 7,
        }
      }
    >
      Currently
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 5,
          "width": "100%",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        maxLength={25}
        onChangeText={[Function]}
        placeholder="Share what youâ€™re up to"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 8,
              "borderWidth": 1,
              "paddingHorizontal": 12,
              "paddingVertical": 4,
            },
            Object {
              "borderBottomWidth": 1,
              "borderWidth": 0,
              "fontSize": 20,
              "marginBottom": 5,
              "paddingLeft": 6,
              "width": "100%",
            },
          ]
        }
        testID="currentlyInput"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "flex-end",
            "height": "100%",
            "opacity": 1,
            "paddingRight": 15,
            "position": "absolute",
            "right": -10,
            "top": 10,
            "width": 70,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../app/assets/ic_remove_text_2.png",
            }
          }
          style={
            Object {
              "height": 15,
              "width": 15,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Inter-Regular",
          "fontSize": 12,
          "marginLeft": 7,
        }
      }
    >
      INVITED BY
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../app/assets/ic_hide.png",
          }
        }
        style={
          Object {
            "height": 17,
          }
        }
      />
    </View>
  </View>
  <View>
    <View
      opacity={0.3}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginTop": 15,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../app/assets/sample-avatar.jpeg",
          }
        }
        style={
          Object {
            "borderRadius": 26,
            "height": 52,
            "marginRight": 11,
            "width": 52,
          }
        }
      />
      <View
        style={
          Object {
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "fontWeight": "500",
            }
          }
        />
        <Text
          style={
            Object {
              "color": "black",
              "fontFamily": "Inter-Regular",
              "fontSize": 10,
            }
          }
        >
          06 December 2021
        </Text>
      </View>
    </View>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "borderColor": "black",
        "borderRadius": 4,
        "borderWidth": 1,
        "height": 43,
        "justifyContent": "center",
        "marginBottom": 20,
        "marginTop": 20,
        "opacity": 1,
        "width": undefined,
      }
    }
    testID="buttonSubmit"
  >
    <Text
      style={
        Object {
          "fontFamily": "Inter-Bold",
          "fontSize": 18,
        }
      }
    >
      I'M DONE
    </Text>
  </View>
</View>
`;
